### DISCUSSION TITLE
Script / guidance to prepare ScaleSet image

### DISCUSSION THREAD

**Discussion by KlausLinzner** [actions/virtual-environments#5843] [General questions]
Hi,
We're running into limits with the hardware resources on the Microsoft Hosted Build Agents in Azure Devops. The documentation at https://docs.microsoft.com/en-us/azure/devops/pipelines/agents/hosted made it pretty clear that there's no way that hardware can be expanded. 

Above referenced documentation also mentions that the alternative for it is to use Azure ScaleSets - which in theory would be a perfect fit. However, it's a huge PITA to configure one of the available images so we can have a rather seamless migration from Microsoft-Hosted to ScaleSet. As I understood from #176 it's not possible for you to provide those images due to legal reasons. 

But while you're not allowed to provide those images directly, is it possible to provide scripts / pipelines to generate the images ourselves as some (all?) of the scripts are already shared in this repo? Like a repo that contains a devops pipeline yml that downloads the base image, applies the configuration and uploads it to a private storage where it can be referenced for the scaleset?
The closest I could find was https://github.com/YannickRe/azuredevops-buildagents which however lacks windows-2022 support and (at least in my environment) is not working well.

If this can't be done either: Would a feature request for devops to specify the used VM be an option? 
F.e. to specify in the pipeline a specific image where the delta in costs to a standard is billed to devops?
Something like the following:
```
pool:
  vmImage: 'windows-2022'
  size: 'Standard_B4ms'
```

BR

**Comment by miketimofeev** [comment]
@KlausLinzner you can build the image yourself and use it wherever you want. Please follow this guide https://github.com/actions/virtual-environments/blob/main/docs/create-image-and-azure-resources.md

**Reply by KlausLinzner** [reply]
First of all thanks for your reply and link - I did not know this one, maybe the link to the doc could be added to above referenced docs site?

And while it helps a lot to get started - can't it just be provided as yml pipeline so it can be used out of the box? Just from a quick look, it's going to take a couple of hours (hopefully, if everything works as expected) until the pipeline is ready, all variables that need configuration are found, all the if/else for windows/linux are added,... 

Unfortunately, devops is really not making it easy for us. The amount of time (and the implicated ongoing maintenance effort) to run on better hardware at the moment is really hard to justify...

**Reply by miketimofeev** [reply]
I'm afraid it's not that easy to provide yml file as configurations vary among different customers. That's why we provide only the basics

**Reply by KlausLinzner** [reply]
But please don't get me wrong: 
You are able to provide one image for all customers worldwide; 
You manage the infrastructure and know of the internals inside out.
If it's not that easy for you - it's nearly impossible for us.

And keep in mind: it's an huge task given to us so we can have f.e. 30 instead of 10GB disc space. 

Does it make sense to discuss this with our MS Sales rep? Can we raise a feature request somewhere?

**Reply by miketimofeev** [reply]
Not sure about feature request as we only provide the scripts to build the images.
You can take a look at our pipelines, which are public here https://github.visualstudio.com/virtual-environments/_build
Yml files are also open-sourced and can be found here https://github.com/actions/virtual-environments/tree/main/images.CI/linux-and-win/azure-pipelines as well as the scripts used in them https://github.com/actions/virtual-environments/tree/main/images.CI/linux-and-win

**Reply by Mikebeem** [reply]
Are those pipelines no longer public? I'm getting: 401 - Uh-oh, you do not have access.
[https://github.visualstudio.com/virtual-environments/_build](https://github.visualstudio.com/virtual-environments/_build)

**Reply by miketimofeev** [reply]
@Mikebeem with the renaming of the main repo the project has also changed its name to https://github.visualstudio.com/runner-images/_build
Sorry for the inconvenience.

**Comment by samus-aran** [comment]
We're locking this discussion because it has not had recent activity and/or other members have asked for more information to assist you but received no response. Thank you for helping us maintain a productive and tidy community for all our members.
